jax/_src/numpy/vectorize.py

```diff
@@ -278,36 +278,43 @@ def wrapped(*args):
         excluded_func, dim_sizes, output_core_dims, error_context)
 
     # Rather than broadcasting all arguments to full broadcast shapes, prefer
-    # expanding dimensions using vmap when possible. By pushing broadcasting
+    # expanding dimensions using vmap. By pushing broadcasting
     # into vmap, we can make use of more efficient batching rules for
     # primitives where only some arguments are batched (e.g., for
-    # lax_linalg.triangular_solve).
+    # lax_linalg.triangular_solve), and avoid instantiating large broadcasted
+    # arrays.
 
-    vec_args = []
-    vmap_counts = []
+    squeezed_args = []
+    rev_filled_shapes = []
 
     for arg, core_dims in zip(args, input_core_dims):
-      # Explicitly broadcast the dimensions already found on each argument,
-      # because these dimensiosns might be of size 1, which vmap doesn't
-      # handle.
-      # TODO(shoyer): Consider squeezing out size 1 dimensions instead, and
-      # doing all vectorization with vmap? This *might* be a little more
-      # efficient but would require more careful book-keeping.
-      core_shape = tuple(dim_sizes[dim] for dim in core_dims)
-      full_shape = broadcast_shape + core_shape
-      vec_shape = full_shape[-arg.ndim:] if arg.ndim else ()
-
-      vec_arg = jnp.broadcast_to(arg, vec_shape)
-      vec_args.append(vec_arg)
-
-      vmap_count = len(vec_shape) - len(core_shape)
-      vmap_counts.append(vmap_count)
+      noncore_shape = arg.shape[:arg.ndim - len(core_dims)]
+
+      pad_ndim = len(broadcast_shape) - len(noncore_shape)
+      filled_shape = pad_ndim * (1,) + noncore_shape
+      assert len(filled_shape) == len(broadcast_shape)
+      assert jnp.broadcast_shapes(broadcast_shape, filled_shape) == broadcast_shape
```

Comment:
If these assertions are intended to catch a user error, we should instead raise an informative error in place of a raw assertion. If these checks are unrelated to user error, we could put them behind `if config.jax_enable_checks`, then the assertions would only be executed during testing, not at normal runtime.

Reply:
They are internal checks, I add `if config.jax_enable_checks`.